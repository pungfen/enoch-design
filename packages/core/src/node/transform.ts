import { compileScript, compileTemplate, parse } from 'vue/compiler-sfc'
import MagicStringBase from 'magic-string'

import type { SFCParseResult } from 'vue/compiler-sfc'
import type { Options } from './plugin'

const DEFINE_FACTORY = 'defineFactory'

class MagicCode extends MagicStringBase {}

const transformCustomBlock = (id: string, sfc: SFCParseResult, options: Options, ms: MagicCode) => {
  if (!sfc.descriptor.customBlocks) return
  const customBlocks = sfc.descriptor.customBlocks
  for (const block of customBlocks) {
    if (block.type === 'pung' && block.lang === 'json') {
      // const blocConfig = JSON.parse(block.content)
      // ms.setConfig(blocConfig)
    }
  }
}

const transformTemplate = (id: string, sfc: SFCParseResult, options: Options, ms: MagicCode) => {
  if (!sfc.descriptor.template) return
}

const transformScript = (id: string, sfc: SFCParseResult, options: Options, ms: MagicCode) => {
  if (sfc.descriptor.scriptSetup) {
    const compiled = compileScript(sfc.descriptor, { id })
    const start = compiled.loc.start
    const end = compiled.loc.end
    let cfg: string = ''
    for (const node of compiled.scriptSetupAst!) {
      switch (node.type) {
        case 'ExpressionStatement':
          const expression = node.expression
          if (expression.type === 'CallExpression' && expression.callee.type === 'Identifier' && expression.callee.name === DEFINE_FACTORY) {
            const [_config, _options] = expression.arguments
            if (_config.type === 'ObjectExpression' && _config.properties) {
              cfg = ms.snip(start.offset + _config.start!, start.offset + _config.end!).toString()
            }
            ms.remove(node.start! + start.offset, node.end! + start.offset)
          }
          break
      }
    }
    if (cfg) {
      ms.prependLeft(
        start.offset,
        `
        \nimport { block as _DF_block, proxy as _DF_proxy } from '@enochfe/core';
        \nconst _DF_BLOCK = _DF_block(${cfg});
        \nconst _DF_PROXY = _DF_proxy(_DF_BLOCK);
      `
      )
    }
  }
}

export const transform = (code: string, id: string, options: Options) => {
  if (!code.includes(DEFINE_FACTORY)) return

  const sfc = parse(code, { filename: id })
  const ms = new MagicCode(code, { filename: id })

  transformCustomBlock(id, sfc, options, ms)
  transformScript(id, sfc, options, ms)
  transformTemplate(id, sfc, options, ms)

  return (
    (ms.hasChanged() && {
      code: ms.toString(),
      get map() {
        return ms.generateMap({
          source: id,
          includeContent: true,
          hires: true
        })
      }
    }) ||
    undefined
  )
}
