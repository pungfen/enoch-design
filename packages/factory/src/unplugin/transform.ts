import { walkAST } from 'ast-walker-scope'
import type { ExportDefaultDeclaration, Statement } from '@babel/types'

import { DEFINE_FACTORY } from './constants'
import { filterMacro, hasPropsOrEmits } from './utils'
import { parseSFC, addNormalScript } from './vue'
import { MagicString } from './magic-string'
import { checkInvalidScopeReference } from './ast'

export const transform = (code: string, id: string) => {
  if (!code.includes(DEFINE_FACTORY)) return

  const sfc = parseSFC(code, id)

  if (!sfc.scriptSetup) return

  const { script, scriptSetup, scriptCompiled } = sfc

  const setupOffset = scriptSetup.loc.start.offset

  // 判断是否存在factory以及factory的重复性
  const nodes = filterMacro(scriptCompiled.scriptSetupAst as Statement[])
  if (nodes.length === 0) return
  else if (nodes.length > 1) throw new SyntaxError(`duplicate ${DEFINE_FACTORY}() call`)

  if (script) checkDefaultExport(scriptCompiled.scriptAst as any)

  const setupBindings = scriptCompiled.scriptSetupAst ? getIdentifiers(scriptCompiled.scriptSetupAst as any) : []

  const s = new MagicString(code)

  const [node] = nodes
  const [arg] = node.arguments

  if (arg) {
    const normalScript = addNormalScript(sfc, s)
    const scriptOffset = normalScript.start()

    s.appendLeft(scriptOffset, `\nimport { defineComponent as _defineComponent } from 'vue';`)
    s.appendLeft(scriptOffset, `\nexport default /*#__PURE__*/ _defineComponent(\n`)

    // if (arg.type === 'ObjectExpression' && hasPropsOrEmits(arg))
    //   throw new SyntaxError(`${DEFINE_FACTORY}() please use defineProps or defineEmits instead.`)

    // checkInvalidScopeReference(arg, DEFINE_FACTORY, setupBindings)

    console.log(s.toString())
    s.moveNode(arg, scriptOffset, { offset: setupOffset })
    console.log(s.toString())

    // removes defineFactory()
    s.remove(setupOffset + node.start!, setupOffset + arg.start!)
    s.remove(setupOffset + arg.end!, setupOffset + node.end!)

    s.appendRight(scriptOffset, '\n);\n')

    normalScript.end()
  } else {
    console.log('debug', s.original)
    // removes defineFactory()
    s.removeNode(node, { offset: setupOffset })
  }

  return getTransformResult(s, id)
}

const getTransformResult = (s: MagicString | undefined, id: string): { code: string; map: any } | undefined => {
  if (s?.hasChanged()) {
    return {
      code: s.toString(),
      get map() {
        return s.generateMap({
          source: id,
          includeContent: true,
          hires: true
        })
      }
    }
  }
}

const checkDefaultExport = (stmts: Statement[]) => {
  const hasDefaultExport = stmts.some((node): node is ExportDefaultDeclaration => node.type === 'ExportDefaultDeclaration')
  if (hasDefaultExport) throw new SyntaxError(`${DEFINE_FACTORY} cannot be used with default export within <script>.`)
}

const getIdentifiers = (stmts: Statement[]) => {
  let ids: string[] = []
  walkAST(
    {
      type: 'Program',
      body: stmts,
      directives: [],
      sourceType: 'module',
      sourceFile: ''
    },
    {
      enter(node) {
        if (node.type === 'BlockStatement') {
          this.skip()
        }
      },
      leave(node) {
        if (node.type !== 'Program') return
        ids = Object.keys(this.scope)
      }
    }
  )

  return ids
}
