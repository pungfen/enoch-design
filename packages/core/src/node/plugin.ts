import { bold, green } from 'kolorist'
import { createUnplugin } from 'unplugin'

import { transform } from './transform'

import type { UnpluginFactory } from 'unplugin'
import type { ResolvedConfig, ViteDevServer } from 'vite'

export interface Options {
  platform?: 'vue' | 'react' | 'weapp'
}

const name = 'unplugin-enochfe'

const unplugin: UnpluginFactory<Options> = (options: Options = {}) => {
  let config: ResolvedConfig

  return {
    name,
    enforce: 'pre',
    vite: {
      configResolved(_config) {
        config = _config
      },
      configureServer(server: ViteDevServer) {
        // const _print = server.printUrls
        // server.printUrls = () => {
        //   const colorUrl = (url: string) => green(url.replace(/:(\d+)\//, (_, port) => `:${bold(port)}/`))
        //   const host = server.resolvedUrls?.local[0] || `${config.server.https ? 'https' : 'http'}://localhost:${config.server.port || '80'}`
        //   _print()
        //   console.log(`  ${green('âžœ')}  ${bold('Inspect')}: ${colorUrl(`${host}__core/`)}`)
        // }
      }
    },

    transform(code, id) {
      try {
        return transform(code, id, options)
      } catch (err: unknown) {
        this.error(`${name} ${err}`)
      }
    }
  }
}

export default createUnplugin(unplugin)
