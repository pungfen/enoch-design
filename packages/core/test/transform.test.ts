import { transform } from '../src/node/transform'

describe('transform', () => {
  const files = import.meta.glob('./*.vue', {
    eager: true,
    as: 'raw'
  })

  for (const [id, code] of Object.entries(files)) {
    test(`transform ${id.replace(/\\/g, '/')}`, () => {
      const exec = () => transform(code, id, {})?.code
      expect(exec()).toMatchSnapshot()
    })
  }
})
