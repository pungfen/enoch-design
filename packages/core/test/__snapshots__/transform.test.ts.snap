// Vitest Snapshot v1

exports[`transform > transform ./demo.vue 1`] = `
"<template>
  <div></div>
</template>

<script setup lang=\\"ts\\">
        
import { block as _DF_block, proxy as _DF_proxy } from '@enochfe/core';
        
const _DF_BLOCK = _DF_block({
  children: {
    form: {
      data: {
        name: 'xx'
      },
      watch: {
        name() {
          console.log('name is change', this.form.data.name)
        }
      },
      children: {
        name: {
          change() {
            console.log(\`I am form's name change\`)
          }
        }
      }
    }
  }
});
        
const _DF_PROXY = _DF_proxy(_DF_BLOCK);
      

</script>
"
`;
